src/sys/socket/mod.rs

```diff
@@ -1631,6 +1641,23 @@ pub fn sendmsg<S>(fd: RawFd, iov: &[IoSlice<'_>], cmsgs: &[ControlMessage],
 }
 
 
+/// Send data in scatter-gather vectors to a socket, possibly accompanied
+/// by ancillary data. Optionally direct the message at the given address,
+/// as with sendto.
+/// sendmsg_pre_alloc is the same as sendmsg but it accepts a preallocated
+/// cmsg buffer vector.
+pub fn sendmsg_pre_alloc<S>(fd: RawFd, iov: &[IoSlice<'_>], cmsgs: &[ControlMessage],
+               flags: MsgFlags, addr: Option<&S>, cmsg_buffer: &mut Vec<u8>) -> Result<usize>
+    where S: SockaddrLike
+{
+    let mhdr = pack_mhdr_to_send(&mut cmsg_buffer[..], iov, cmsgs, addr);
```

Comment:
We need to ensure that `cmsg_buffer` has enough length to accommodate `cmsgs` before passing it to `pack_mhdr_to_send()`

Reply:
The current impl of `pack_mhdr_to_send()` requires that `cmsg_buffer.len()` equals the length needed by control messages, which may not be true with the pre-alloc API, we need to change this:

https://github.com/nix-rust/nix/blob/7d75bbc9d6cfe91b903c91f39ed8b500ec65b8d1/src/sys/socket/mod.rs#L2044

Reply:
Maybe we can add a if-condition in `sendmsg_pre_alloc()` which will panic if `cmsg_buffer.len() != cmsgs .len()`?

Reply:
I kinda think that would be too limited? There should be no issue if `cmsg_buffer.len()`> `cmsg.len()`, my preferred way is to: check if `cmsg_buffer.len()` is greater than or equal to `cmsg.len()`, if not, return `Err(Errno::ENOBUFS)`. For the `pack_mhdr_to_send()` function, we can add another `cmsg_len` argument, and document that `cmsg_buffer.len()` could be greater than `cmsg_len` due to the this `sendmsg_pralloc()` function.